### DISCUSSION TITLE
How to set tag value of date type?

### DISCUSSION THREAD

**Discussion by mateusz-glowinski** [dmroeder/pylogix#220] [General]
I've noticed that `Date/Time` tag type is returned by `GetTagList` as `UDINT`.

Can I somehow detect that it's a tag with `Date/Time` type?

How should i set a value of such a tag using `pylogix`?

**Comment by TheFern2** [comment]
I want to say that's just an epoch value. So convert your date time to epoch and then write to tag.

For figuring out if is a date/time value tag, I don't know if there's an attribute in the Tag class but you could check whether the value is epoch valid. i.e. if the value is up to date meaning within a month, you could check whether the epoch value is greater than N day. This could disqualify other udint values that aren't dates/times. Another way is to give proper tag names to datetime tags.

**Reply by mateusz-glowinski** [reply]
Sorry, i didn't get it.
Do you mean something like this: https://stackoverflow.com/questions/12400256/converting-epoch-time-into-the-datetime?

**Reply by TheFern2** [reply]
I just looked at the types screenshot and this might not work. I was thinking the value was an epoch value but it looks like there's two separate tag types one for date and one for time. I have no clue how date and time types work. What does the 3 remark says about it on your screenshot?

**Reply by mateusz-glowinski** [reply]
`3. Can be used by sending data to UDINT, mainly for us e with PanelView Plus and PanelView 800 HMI terminals.`

**Reply by mateusz-glowinski** [reply]
Unfortunately, there is no information about format, in which the information is stored.
I've found only this:
![image](https://user-images.githubusercontent.com/31914443/221599823-d3d97c3f-02d4-40e1-9669-1835549e1a87.png)

**Reply by TheFern2** [reply]
Ah ok, based on this screenshot at least DATE is epoch for sure, so you can try feeding date in epochtime in seconds. TIME doesn't say whether is epoch though, does that mean is just milliseconds for the current hour? Maybe try epoch there too, but I am not 100% sure on that one. I wonder if plctalk has some post on this.

**Reply by mateusz-glowinski** [reply]
Ok, i will check that out. Thanks for a tip :)

**Reply by TheFern2** [reply]
No problem, yeah sorry couldn't be more definitive, but I don't remember ever dealing with these types in any of our internal projects. Surely someone has, but if it only works with those ab panels, I'd probably stay away from those types and use something more generic.

**Reply by dmroeder** [reply]
It appears that the documentation is from the Micro800 platform.  You can choose a DATE type and a TIME type, both of which are 4 byte unsigned integers.  When requesting the tag database, TIME and DATE types all report 0xc8 (UDINT).  Neither appear to be unique data types, kind of strange.

**Reply by TheFern2** [reply]
yeah if is udint is hard to know unless it is using epoch then you can sorta validate, I would imagine it is, how else would time be calculated? If epoch is out of the question, I would just append all my tags with _time, and _date. However if plain epoch can fit in an unsigned type I would just use epoch instead and avoid all this mystery with this special types.

**Reply by mateusz-glowinski** [reply]
I will try out this epoch solution. Once i check that out, i will let you know.

**Reply by TheFern2** [reply]
One thing to note is that to represent the entire epoch you need a floating type. I am not 100% positive if epoch fits in REAL, or LREAL. If epoch is larger than the max size of these types. Then you need to split secs and nsecs kind of like is done in the ROS framework Time class http://wiki.ros.org/roscpp/Overview/Time link for reference but you'll need a struct in the PLC with two int32 types. Also well epoch won't fit on 32bit systems by 2038 as that number is growing daily https://en.wikipedia.org/wiki/Year_2038_problem.

**Reply by kyle-github** [reply]
Most POSIX systems moved to 64-bit signed integers for epoch time in seconds.   That's enough for a while :-)  LINT on a AB PLC.

**Comment by dmroeder** [comment]
I'm unaware of a date/time tag type.  There is display style of date/time for LINT types.  Is this what you are referring to?

**Reply by mateusz-glowinski** [reply]
Yes, exactly. 

![image](https://user-images.githubusercontent.com/31914443/221503447-0230f497-bbb8-4a89-bc74-04bf94d1d22e.png)


My intention is to detect, that variable should be displayed as date/time :)

**Comment by mateusz-glowinski** [comment]
@TheFern2 
Indeed, converting and parsing `DATE` value to epoch value did a job. Thanks a lot!

**Reply by TheFern2** [reply]
Nice! Glad to hear that!
